src/Plugin/QueueWorker/EloquaAppCloudQueueWorkerBase.php

```diff
@@ -0,0 +1,186 @@
+        try {
+
+          // Upload content to Eloqua staging area.
+          $bulkApi->upload($contactsToBeStaged);
+          // Uploaded contacts successfully.
+          return;
+        } catch (Exception $e) {
+          $msg = "Reupload attempt #$retries Failed! Caught Exception: " . $e->getMessage();
+          $logger->error($msg);
+        }
+      }
+
+      $msg = "Upload failed after multiple retries.";
+      $msg .= print_r($bulkApi->log(), TRUE);
+      $msg .= "Caught Exception: " . $e->getMessage();
+      $logger->error($msg);
+    }
+  }
+
+  /**
+   * Wraps try/catch and retry attempts around Elomentary's $bulkApi->sync()
+   * function.
+   *
+   * @param {Object} $bulkApi
+   *    Elomentary bulkApi object reference.
+   *
+   * @param {Logger} $logger
+   *    Logger object for logging routine info / errors.
+   *
+   * @return {String} $uri
+   *    The Sync endpoint URI used to access Eloqua Sync data.
+   */
+  protected function tryBulkApiSync($bulkApi, LoggerInterface $logger) {
+    $retries = 1;
+    // In case of API failures, up to 5 retry attempts.
+    $maxRetries = 5;
+
+    try {
+      // Sync uploaded contacts in staging with Eloqua.
+      $bulkApi->sync();
+      $currentStatus = $this->getBulkApiStatus($bulkApi);
+
+      // $bulkApi->status() when passed true,
+      // will block until the upload succeeds or fails.
+      if ($currentStatus === 'success' || $currentStatus === 'warning') {
+        $msg = "Status Returned: " . $currentStatus;
+        $logger->info($msg);
+        $msg = "Sync completed successfully!";
+        $logger->info($msg);
+      }
+
+      $syncResponse = $bulkApi->getResponse('sync', NULL);
+
+      $msg = "syncResponse is: " . print_r($syncResponse, TRUE) . "\n";
+      $logger->info($msg);
+      $uri = trim($syncResponse['uri'], '/');
+      $msg = "Import URI is: $uri\n";
+      $logger->info($msg);
+      return $uri;
+
+    } catch (Exception $e) {
+      $msg = "Sync Failed! Caught Exception: " . $e->getMessage();
+      $logger->error($msg);
+
+      for ($retries; $retries <= $maxRetries; $retries++) {
+        try {
+          // Attempt to sync.
+          $bulkApi->sync();
+
+          // Poll for sync status response.
+          $currentStatus = $this->getBulkApiStatus($bulkApi);
+
+          // A successful status response is either 'success' or 'warning'.
+          if ($currentStatus === 'success' || $currentStatus === 'warning') {
+            $msg = "Status Returned: " . $currentStatus;
+            $logger->info($msg);
+            $msg = "Sync completed successfully!";
+            $logger->info($msg);
+            $syncResponse = $bulkApi->getResponse('sync', NULL);
+            $uri = trim($syncResponse['uri'], '/');
+            $logger->info($msg);
+            return $uri;
+            return $uri;
+          }
+
+        } catch (Exception $e) {
+          $msg = "Sync retry attempt #$retries returned status: $currentStatus.";
```

Comment:
Same deal here. Do we want to throw our own exception so we can handle the failure?

Reply:
Ditto.